File: app/src/main/java/org/y20k/transistor/adapter/CollectionViewModel.java
/**
 * CollectionViewModel.java
 * Implements the CollectionViewModel class
 * A CollectionViewModel stores the Collection of stations in a ViewModel
 *
 * This file is part of
 * TRANSISTOR - Radio App for Android
 *
 * Copyright (c) 2015-17 - Y20K.org
 * Licensed under the MIT-License
 * http://opensource.org/licenses/MIT
 */

package org.y20k.transistor.adapter;

import android.app.Application;
import android.arch.lifecycle.AndroidViewModel;
import android.arch.lifecycle.MutableLiveData;
import android.content.Context;
import android.content.SharedPreferences;
import android.os.AsyncTask;
import android.preference.PreferenceManager;

import org.y20k.transistor.core.Station;
import org.y20k.transistor.helpers.LogHelper;
import org.y20k.transistor.helpers.StationListHelper;
import org.y20k.transistor.helpers.TransistorKeys;

import java.util.ArrayList;

/**
 * CollectionViewModel.class
 */
public class CollectionViewModel extends AndroidViewModel implements TransistorKeys {

    /* Define log tag */
    private static final String LOG_TAG = CollectionViewModel.class.getSimpleName();


    /* Main class variables */
    private MutableLiveData<ArrayList<Station>> mStationListLiveData;
    private MutableLiveData<Station> mPlayerServiceStationLiveData;
    private MutableLiveData<Boolean> mTwoPaneLiveData;

    /* Constructor */
    public CollectionViewModel(Application application) {
        super(application);

        // initialize LiveData
        mStationListLiveData = new MutableLiveData<ArrayList<Station>>();
        mPlayerServiceStationLiveData = new MutableLiveData<Station>();
        mTwoPaneLiveData = new MutableLiveData<Boolean>();

        // load state from shared preferences and set live data values
        loadAppState(application);

        // set station from PlayerService to null
        mPlayerServiceStationLiveData.setValue(null);

        // load station list from storage and set live data
        mStationListLiveData.setValue(StationListHelper.loadStationListFromStorage(application));

        // load station list and set live data in background -> not used because DiffResult.dispatchUpdatesTo is causing problems in Adapter
        // new LoadCollectionAsyncTask().execute(application);
    }


    @Override
    protected void onCleared() {
        super.onCleared();
    }


    /* Loads app state from preferences and updates live data */
    private void loadAppState(Context context) {
        SharedPreferences settings = PreferenceManager.getDefaultSharedPreferences(context);
        mTwoPaneLiveData.setValue(settings.getBoolean(PREF_TWO_PANE, false));
        LogHelper.v(LOG_TAG, "Loading state and updating live data.");
    }


    /* Getter for StationList */
    public MutableLiveData<ArrayList<Station>> getStationList() {
        return mStationListLiveData;
    }


    /* Getter for station from PlayerService */
    public MutableLiveData<Station> getPlayerServiceStation() {
        return mPlayerServiceStationLiveData;
    }


    /* Getter for TwoPane */
    public MutableLiveData<Boolean> getTwoPane() {
        return mTwoPaneLiveData;
    }


    /**
     * Inner class: AsyncTask that loads list in background
     */
    class LoadCollectionAsyncTask extends AsyncTask<Context, Void, ArrayList<Station>> {
        @Override
        protected ArrayList<Station> doInBackground(Context... contexts) {
//            // stress test for DiffResult todo remove
//            try {
//                Thread.sleep(1500);
//            } catch (InterruptedException e) {
//                e.printStackTrace();
//            }

            // load stations in background
            return StationListHelper.loadStationListFromStorage(contexts[0]);
        }

        @Override
        protected void onPostExecute(ArrayList<Station> stationList) {
            // set live data
            mStationListLiveData.setValue(stationList);
        }
    }
    /**
     * End of inner class
     */

}

File: app/src/main/java/org/y20k/transistor/collection/CollectionViewModel.kt
/*
 * CollectionViewModel.kt
 * Implements the CollectionViewModel class
 * A CollectionViewModel stores the collection of stations as live data
 *
 * This file is part of
 * TRANSISTOR - Radio App for Android
 *
 * Copyright (c) 2015-22 - Y20K.org
 * Licensed under the MIT-License
 * http://opensource.org/licenses/MIT
 */


package org.y20k.transistor.collection

import android.app.Application
import android.content.BroadcastReceiver
import android.content.Context
import android.content.Intent
import android.content.IntentFilter
import androidx.lifecycle.AndroidViewModel
import androidx.lifecycle.MutableLiveData
import androidx.lifecycle.viewModelScope
import androidx.localbroadcastmanager.content.LocalBroadcastManager
import kotlinx.coroutines.launch
import org.y20k.transistor.Keys
import org.y20k.transistor.core.Collection
import org.y20k.transistor.helpers.FileHelper
import org.y20k.transistor.helpers.LogHelper
import java.util.*


/*
 * CollectionViewModel.class
 */
class CollectionViewModel(application: Application) : AndroidViewModel(application) {

    /* Define log tag */
    private val TAG: String = LogHelper.makeLogTag(CollectionViewModel::class.java)


    /* Main class variables */
    val collectionLiveData: MutableLiveData<Collection> = MutableLiveData<Collection>()
    val collectionSizeLiveData: MutableLiveData<Int> = MutableLiveData<Int>()
    private var modificationDateViewModel: Date = Date()
    private var collectionChangedReceiver: BroadcastReceiver


    /* Init constructor */
    init {
        // load collection
        loadCollection(application)
        // create and register collection changed receiver
        collectionChangedReceiver = createCollectionChangedReceiver()
        LocalBroadcastManager.getInstance(application).registerReceiver(collectionChangedReceiver, IntentFilter(Keys.ACTION_COLLECTION_CHANGED))
    }


    /* Overrides onCleared */
    override fun onCleared() {
        super.onCleared()
        LocalBroadcastManager.getInstance(getApplication()).unregisterReceiver(collectionChangedReceiver)
    }


    /* Creates the collectionChangedReceiver - handles Keys.ACTION_COLLECTION_CHANGED */
    private fun createCollectionChangedReceiver(): BroadcastReceiver {
        return object : BroadcastReceiver() {
            override fun onReceive(context: Context, intent: Intent) {
                if (intent.hasExtra(Keys.EXTRA_COLLECTION_MODIFICATION_DATE)) {
                    val date: Date = Date(intent.getLongExtra(Keys.EXTRA_COLLECTION_MODIFICATION_DATE, 0L))
                    // check if reload is necessary
                    if (date.after(modificationDateViewModel)) {
                        LogHelper.v(TAG, "CollectionViewModel - reload collection after broadcast received.")
                        loadCollection(context)
                    }
                }
            }
        }
    }


    /* Reads collection of radio stations from storage using GSON */
    private fun loadCollection(context: Context) {
        LogHelper.v(TAG, "Loading collection of stations from storage")
        viewModelScope.launch {
            // load collection on background thread
            val collection: Collection = FileHelper.readCollectionSuspended(getApplication())
            // get updated modification date
            modificationDateViewModel = collection.modificationDate
            // update collection view model
            collectionLiveData.value = collection
            // update collection sie
            collectionSizeLiveData.value = collection.stations.size
        }
    }

}
